### ISSUE TITLE
union does not produce expeced result when endpoints of a polygon overlap

### ISSUE THREAD

**Issue by bacchanalia** [diagrams/diagrams-contrib#77]
In the following code ex4 unexpectedly produces an empty diagram.
```haskell
{-# LANGUAGE TypeFamilies, FlexibleContexts #-}
import Diagrams.Prelude
import Diagrams.Backend.Cairo.CmdLine
import Diagrams.TwoD.Polygons
import qualified Diagrams.TwoD.Path.Boolean as B

rose n r1 r2 = polygon $ with
  & polyType   .~ PolyPolar (replicate n $ 1/8 @@ turn) (cycle [r1, r2])
  & polyOrient .~ NoOrient

roseWithNoOverlap = rose 7
roseWithOverlap   = rose 8

roseWithNoOverlapOrthoginal = roseWithNoOverlap 3 1
roseWithNoOverlapDiagonal   = roseWithNoOverlap 1 3
roseWithOverlapOrthoginal   = roseWithOverlap   3 1
roseWithOverlapDiagonal     = roseWithOverlap   1 3

ex d = stroke $ B.union Winding $ d <> circle 2

ex1, ex2, ex3, ex4, dia :: Diagram Cairo
ex1 = ex roseWithNoOverlapOrthoginal  -- works
ex2 = ex roseWithNoOverlapDiagonal    -- works: equiv to (ex1 # rotate (1/8 @@ turn))
ex3 = ex roseWithOverlapOrthoginal    -- works: equiv to ex1
ex4 = ex roseWithOverlapDiagonal      -- broken: produces nothing, expected quiv to ex2

dia = (ex1 ||| ex2) === (ex3 ||| ex4)
main = mainWith dia
```

**Comment by byorgey** [comment]
This may be related to https://github.com/kuribas/cubicbezier/issues/6 .  I know there was another similar example where reversing one of the paths involved caused it to work --- does this example exhibit the same behavior?

**Comment by bacchanalia** [comment]
neither d or (circle 2) reversed
![neither](https://user-images.githubusercontent.com/852852/54061304-43ea2e80-41ce-11e9-992f-14771ec9c8b3.png)
d reversed
![d](https://user-images.githubusercontent.com/852852/54061308-49477900-41ce-11e9-95bd-34d8f1d4b22f.png)
(circle 2) reversed
![circle](https://user-images.githubusercontent.com/852852/54061313-4d739680-41ce-11e9-9ca7-3f017e5911b4.png)
both d and (circle 2) reversed
![both](https://user-images.githubusercontent.com/852852/54061319-51071d80-41ce-11e9-8988-44c5a2805703.png)

**Comment by byorgey** [comment]
Weird!

**Comment by fryguybob** [comment]
The way paths are converted to be used with `cubicbezier` package leads to a tiny segment.  I think this tiny segment (`straight (V2 (-0.0) 4.440892098500626e-16)`) is not handled well by `cubicbezier`.  Boolean operations on non-convex polygons is super hard to get right in all the strange cases.  This only gets harder with cubic segments.  We could filter out tiny segments before handing off to `cubicbezier`.  

What was the reason you switched to the "overlap" version again?

**Comment by bacchanalia** [comment]
I didn't switch back to the overlap version. I just figured: see weird behaviour, post weird behaviour.

**Comment by fryguybob** [comment]
> I just figured: see weird behaviour, post weird behaviour.

Absolutely!  I just was wondering if there was a need for that particular overlapped form in your workflow.
